[illegal_argument_exception] cannot set [search_type] when using [knn] search

dependencies：
org.springframework.boot:spring-boot-starter-data-elasticsearch:3.1.0
org.springframework.boot;spring-boot-starter:3.1.0

search code：

`public List<RetrieverDoc> queryByEmbedding(String queryString, int topK) throws Exception {`
    `    float[] predict = textFeatureExtractionService.predict(queryString, true);`
    `    List<Float> vector = new ArrayList<>();`
    `    for (float v : predict) {`
    `        vector.add(v);`
    `    }`
    `    KnnQuery.Builder knnBuilder = new KnnQuery.Builder();`
    `    knnBuilder.field("embedding");`
    `    knnBuilder.queryVector(vector);`
    `    knnBuilder.numCandidates(topK);`
    `    knnBuilder.k(topK);`
    `    NativeQuery nativeQuery = NativeQuery.builder()`
    `            .withKnnQuery(knnBuilder.build())`
    `            .build();`
    `    SearchHits<RetrieverDoc> search = elasticsearchOperations.search(nativeQuery, RetrieverDoc.class);`
    `    List<RetrieverDoc> result = new ArrayList<>();`
    `    for (SearchHit<RetrieverDoc> retrieverDocSearchHit : search) {`
    `        RetrieverDoc content = retrieverDocSearchHit.getContent();`
    `        content.setScore(retrieverDocSearchHit.getScore());`
    `        result.add(content);`
    `    }`
    `    return result;`
    `}`

exception ：


`[es/search] failed: [illegal_argument_exception] cannot set [search_type] when using [knn] search, since the search type is determined automatically`
`org.springframework.data.elasticsearch.UncategorizedElasticsearchException: [es/search] failed: [illegal_argument_exception] cannot set [search_type] when using [knn] search, since the search type is determined automatically`
`	at app//org.springframework.data.elasticsearch.client.elc.ElasticsearchExceptionTranslator.translateExceptionIfPossible(ElasticsearchExceptionTranslator.java:102)`
`	at app//org.springframework.data.elasticsearch.client.elc.ElasticsearchExceptionTranslator.translateException(ElasticsearchExceptionTranslator.java:63)`
`	at app//org.springframework.data.elasticsearch.client.elc.ElasticsearchTemplate.execute(ElasticsearchTemplate.java:625)`
`	at app//org.springframework.data.elasticsearch.client.elc.ElasticsearchTemplate.doSearch(ElasticsearchTemplate.java:341)`
`	at app//org.springframework.data.elasticsearch.client.elc.ElasticsearchTemplate.search(ElasticsearchTemplate.java:334)`
`	at app//org.springframework.data.elasticsearch.core.AbstractElasticsearchTemplate.search(AbstractElasticsearchTemplate.java:492)`
`	at app//com.longrise.nlp.service.ElasticsearchService.queryByEmbedding(ElasticsearchService.java:64)`
`	at app//com.longrise.nlp.NlpApplicationTests.testQueryDoc(NlpApplicationTests.java:52)`
`	at java.base@17.0.7/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)`
`	at java.base@17.0.7/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)`
`	at java.base@17.0.7/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)`
`	at java.base@17.0.7/java.lang.reflect.Method.invoke(Method.java:568)`
`	at app//org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)`
`	at app//org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)`
`	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)`
`	at app//org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)`
`	at app//org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)`
`	at app//org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)`
`	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)`
`	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)`
`	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)`
`	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)`
`	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)`
`	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)`
`	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)`
`	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)`
`	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:217)`
`	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)`
`	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)`
`	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)`
`	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)`
`	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)`
`	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)`
`	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)`
`	at java.base@17.0.7/java.util.ArrayList.forEach(ArrayList.java:1511)`
`	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)`
`	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)`
`	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)`
`	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)`
`	at java.base@17.0.7/java.util.ArrayList.forEach(ArrayList.java:1511)`
`	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)`
`	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)`
`	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)`
`	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)`
`	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)`
`	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)`
`	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)`
`	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)`
`	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)`
`	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)`
`	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)`
`	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)`
`	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)`
`	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)`
`	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)`
`	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)`
`	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)`
`	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)`
`	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)`
`	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)`
`	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:62)`
`	at java.base@17.0.7/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)`
`	at java.base@17.0.7/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)`
`	at java.base@17.0.7/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)`
`	at java.base@17.0.7/java.lang.reflect.Method.invoke(Method.java:568)`
`	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)`
`	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)`
`	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)`
`	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)`
`	at jdk.proxy1/jdk.proxy1.$Proxy2.stop(Unknown Source)`
`	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)`
`	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)`
`	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)`
`	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)`
`	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)`
`	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)`
`	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)`
`	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)`
`	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)`
`Caused by: co.elastic.clients.elasticsearch._types.ElasticsearchException: [es/search] failed: [illegal_argument_exception] cannot set [search_type] when using [knn] search, since the search type is determined automatically`
`	at app//co.elastic.clients.transport.rest_client.RestClientTransport.getHighLevelResponse(RestClientTransport.java:334)`
`	at app//co.elastic.clients.transport.rest_client.RestClientTransport.performRequest(RestClientTransport.java:154)`
`	at app//co.elastic.clients.elasticsearch.ElasticsearchClient.search(ElasticsearchClient.java:1882)`
`	at app//org.springframework.data.elasticsearch.client.elc.ElasticsearchTemplate.lambda$doSearch$14(ElasticsearchTemplate.java:341)`
`	at app//org.springframework.data.elasticsearch.client.elc.ElasticsearchTemplate.execute(ElasticsearchTemplate.java:623)`
`	... 90 more`

--- a/src/main/java/org/springframework/data/elasticsearch/client/elc/RequestConverter.java
+++ b/src/main/java/org/springframework/data/elasticsearch/client/elc/RequestConverter.java
@@ -22,6 +22,7 @@ import co.elastic.clients.elasticsearch._types.Conflicts;
 import co.elastic.clients.elasticsearch._types.FieldValue;
 import co.elastic.clients.elasticsearch._types.InlineScript;
 import co.elastic.clients.elasticsearch._types.OpType;
+import co.elastic.clients.elasticsearch._types.SearchType;
 import co.elastic.clients.elasticsearch._types.SortOptions;
 import co.elastic.clients.elasticsearch._types.SortOrder;
 import co.elastic.clients.elasticsearch._types.VersionType;
@@ -1153,9 +1154,12 @@ class RequestConverter {
 				var query = param.query();
 				mrb.searches(sb -> sb //
 						.header(h -> {
+							var searchType = (query instanceof NativeQuery nativeQuery && nativeQuery.getKnnQuery() != null) ? null
+									: searchType(query.getSearchType());
+
 							h //
 									.index(Arrays.asList(param.index().getIndexNames())) //
-									.searchType(searchType(query.getSearchType())) //
+									.searchType(searchType) //
 									.requestCache(query.getRequestCache()) //
 							;
 
@@ -1256,8 +1260,8 @@ class RequestConverter {
 							query.getScriptedFields().forEach(scriptedField -> bb.scriptFields(scriptedField.getFieldName(),
 									sf -> sf.script(getScript(scriptedField.getScriptData()))));
 
-							if (query instanceof NativeQuery) {
-								prepareNativeSearch((NativeQuery) query, bb);
+							if (query instanceof NativeQuery nativeQuery) {
+								prepareNativeSearch(nativeQuery, bb);
 							}
 							return bb;
 						} //
@@ -1279,12 +1283,15 @@ class RequestConverter {
 
 		ElasticsearchPersistentEntity<?> persistentEntity = getPersistentEntity(clazz);
 
+		var searchType = (query instanceof NativeQuery nativeQuery && nativeQuery.getKnnQuery() != null) ? null
+				: searchType(query.getSearchType());
+
 		builder //
 				.version(true) //
 				.trackScores(query.getTrackScores()) //
 				.allowNoIndices(query.getAllowNoIndices()) //
 				.source(getSourceConfig(query)) //
-				.searchType(searchType(query.getSearchType())) //
+				.searchType(searchType) //
 				.timeout(timeStringMs(query.getTimeout())) //
 				.requestCache(query.getRequestCache()) //
 		;
@@ -1361,8 +1368,8 @@ class RequestConverter {
 		query.getScriptedFields().forEach(scriptedField -> builder.scriptFields(scriptedField.getFieldName(),
 				sf -> sf.script(getScript(scriptedField.getScriptData()))));
 
-		if (query instanceof NativeQuery) {
-			prepareNativeSearch((NativeQuery) query, builder);
+		if (query instanceof NativeQuery nativeQuery) {
+			prepareNativeSearch(nativeQuery, builder);
 		}
 
 		if (query.getTrackTotalHits() != null) {

--- a/src/main/java/org/springframework/data/elasticsearch/core/query/BaseQuery.java
+++ b/src/main/java/org/springframework/data/elasticsearch/core/query/BaseQuery.java
@@ -58,7 +58,7 @@ public class BaseQuery implements Query {
 	protected float minScore;
 	@Nullable protected Collection<String> ids;
 	@Nullable protected String route;
-	protected SearchType searchType = SearchType.QUERY_THEN_FETCH;
+	@Nullable protected SearchType searchType = SearchType.QUERY_THEN_FETCH;
 	@Nullable protected IndicesOptions indicesOptions;
 	protected boolean trackScores;
 	@Nullable protected String preference;
@@ -278,10 +278,11 @@ public class BaseQuery implements Query {
 		this.route = route;
 	}
 
-	public void setSearchType(SearchType searchType) {
+	public void setSearchType(@Nullable SearchType searchType) {
 		this.searchType = searchType;
 	}
 
+	@Nullable
 	@Override
 	public SearchType getSearchType() {
 		return searchType;

--- a/src/main/java/org/springframework/data/elasticsearch/core/query/BaseQueryBuilder.java
+++ b/src/main/java/org/springframework/data/elasticsearch/core/query/BaseQueryBuilder.java
@@ -46,7 +46,7 @@ public abstract class BaseQueryBuilder<Q extends BaseQuery, SELF extends BaseQue
 	private float minScore;
 	private final Collection<String> ids = new ArrayList<>();
 	@Nullable private String route;
-	private Query.SearchType searchType = Query.SearchType.QUERY_THEN_FETCH;
+	@Nullable private Query.SearchType searchType = Query.SearchType.QUERY_THEN_FETCH;
 	@Nullable private IndicesOptions indicesOptions;
 	private boolean trackScores;
 	@Nullable private String preference;
@@ -140,6 +140,7 @@ public abstract class BaseQueryBuilder<Q extends BaseQuery, SELF extends BaseQue
 		return indicesBoost;
 	}
 
+	@Nullable
 	public Query.SearchType getSearchType() {
 		return searchType;
 	}
@@ -250,11 +251,11 @@ public abstract class BaseQueryBuilder<Q extends BaseQuery, SELF extends BaseQue
 		return self();
 	}
 
-		/**
-		 * Set Ids for a multi-get request run with this query. Not used in any other searches.
-		 *
-		 * @param ids list of id values
-		 */
+	/**
+	 * Set Ids for a multi-get request run with this query. Not used in any other searches.
+	 *
+	 * @param ids list of id values
+	 */
 	public SELF withIds(String... ids) {
 
 		this.ids.clear();
@@ -262,11 +263,11 @@ public abstract class BaseQueryBuilder<Q extends BaseQuery, SELF extends BaseQue
 		return self();
 	}
 
-		/**
-		 * Set Ids for a multi-get request run with this query. Not used in any other searches.
-		 *
-		 * @param ids list of id values
-		 */
+	/**
+	 * Set Ids for a multi-get request run with this query. Not used in any other searches.
+	 *
+	 * @param ids list of id values
+	 */
 	public SELF withIds(Collection<String> ids) {
 
 		Assert.notNull(ids, "ids must not be null");
@@ -342,7 +343,7 @@ public abstract class BaseQueryBuilder<Q extends BaseQuery, SELF extends BaseQue
 		return self();
 	}
 
-	public SELF withSearchType(Query.SearchType searchType) {
+	public SELF withSearchType(@Nullable Query.SearchType searchType) {
 		this.searchType = searchType;
 		return self();
 	}
@@ -382,12 +383,12 @@ public abstract class BaseQueryBuilder<Q extends BaseQuery, SELF extends BaseQue
 		return self();
 	}
 
-		/**
-		 * Set Ids with routing values for a multi-get request run with this query. Not used in any other searches.
-		 *
-		 * @param idsWithRouting list of id values, must not be {@literal null}
-		 * @since 4.3
-		 */
+	/**
+	 * Set Ids with routing values for a multi-get request run with this query. Not used in any other searches.
+	 *
+	 * @param idsWithRouting list of id values, must not be {@literal null}
+	 * @since 4.3
+	 */
 	public SELF withIdsWithRouting(List<Query.IdWithRouting> idsWithRouting) {
 
 		Assert.notNull(idsWithRouting, "idsWithRouting must not be null");

--- a/src/main/java/org/springframework/data/elasticsearch/core/query/Query.java
+++ b/src/main/java/org/springframework/data/elasticsearch/core/query/Query.java
@@ -221,6 +221,7 @@ public interface Query {
 	 *
 	 * @return
 	 */
+	@Nullable
 	SearchType getSearchType();
 
 	/**

